docs/manuals/man1/PrettyPrintCrl.1.md
# PrettyPrintCrl 1 "July 20, 2016" PKI "PKI CRL Print Tool"

## NAME

PrettyPrintCrl - reads a certificate revocation list (CRL) stored in an ASCII base-64 encoded file
and outputs it in a readable format.

## SYNOPSIS

**PrettyPrintCrl** *input-file* [*output-file*]

## DESCRIPTION

The **PrettyPrintCrl** command provides a command-line utility used to print the contents of a CRL
stored as ASCII base-64 encoded data in a file to a readable format.
The output of this command is displayed to standard output,
but can be optionally saved into a specified file.

## OPTIONS

**&lt;input-file&gt;**  
    Mandatory. Specifies the path to the file that contains the ASCII base-64 encoded CRL.

**&lt;output-file&gt;**  
    Optional. Specifies the path to the file to write the CRL.
    If the output file is not specified, the CRL information is written to the standard output.

## EXAMPLES

The following example **PrettyPrintCrl** command takes the ASCII base-64 encoded CRL
in the ascii_data.crl file and writes the CRL in the pretty-print format
to the output file crl.out:

```
$ PrettyPrintCrl ascii_data.crl crl.out
```

For this example, the base-64 encoded CRL data in the ascii_data.crl looks like the following:

```
-----BEGIN X509 CRL-----
MIICVDCCATwCAQEwDQYJKoZIhvcNAQELBQAwTjErMCkGA1UECgwidXNlcnN5cy5y
ZWRoYXQuY29tIFNlY3VyaXR5IERvbWFpbjEfMB0GA1UEAwwWQ0EgU2lnbmluZyBD
ZXJ0aWZpY2F0ZRcNMTYwNzIyMjExMjUwWhcNMTYwNzIyMjMwMDAwWjCBiDAgAgEK
Fw0xNjA3MjIyMDU1MTZaMAwwCgYDVR0VBAMKAQYwIAIBCRcNMTYwNzIyMjEwMTU2
WjAMMAoGA1UdFQQDCgEGMCACAQgXDTE2MDcyMjIxMTIyNVowDDAKBgNVHRUEAwoB
ATAgAgEHFw0xNjA3MjIyMTAxNTZaMAwwCgYDVR0VBAMKAQagLzAtMB8GA1UdIwQY
MBaAFLs2mF1ly4jghyM3b1v3r4uK67q1MAoGA1UdFAQDAgEKMA0GCSqGSIb3DQEB
CwUAA4IBAQCjnwpdLVU4sg3GnOFQiHpBuWspevzj0poHQs9b4Uv17o0MC4irftkR
zRBVgwLvdSd5WFEUSbhWVjhS4o4w84BXdmti/+UBS+mOVNxiKqs3Z7Fxcg+mCsiH
SDWT3iiqZVqlPMOKDzIQGj4XeArSBK13qjNdwKzVJZlXYfwzdDtyVKBJcoETXGZ3
irU8RTXo7OhO6xKDAaHjzVVynjfGdIDaavl1fjwXFufwZBeiXm1zyyFSvDUdny4G
29NTmM2945jCESeR7DV2q1LHG/v2rzCOKTWdPdXTPCics05KzUA4S6X+mp051wkh
yJM2LYpV6lKV6JiczHLrgf5QcqfwSkTX
-----END X509 CRL-----
```

The CRL in pretty-print format in the crl.out file looks like the following:

```
    Certificate Revocation List:
        Data:
            Version:  v2
            Signature Algorithm: SHA256withRSA - 1.2.840.113549.1.1.11
            Issuer: CN=CA Signing Certificate,O=example.com Security Domain
            This Update: Friday, July 22, 2016 3:12:50 PM MDT America/Denver
            Next Update: Friday, July 22, 2016 5:00:00 PM MDT America/Denver
            Revoked Certificates:
                Serial Number: 0xA
                Revocation Date: Friday, July 22, 2016 2:55:16 PM MDT America/Denver
                Extensions:
                    Identifier: Revocation Reason - 2.5.29.21
                        Critical: no
                        Reason: CA_Compromise
                Serial Number: 0x9
                Revocation Date: Friday, July 22, 2016 3:01:56 PM MDT America/Denver
                Extensions:
                    Identifier: Revocation Reason - 2.5.29.21
                        Critical: no
                        Reason: Affiliation_Changed
                Serial Number: 0x8
                Revocation Date: Friday, July 22, 2016 3:12:25 PM MDT America/Denver
                Extensions:
                    Identifier: Revocation Reason - 2.5.29.21
                        Critical: no
                        Reason: Key_Compromise
                Serial Number: 0x7
                Revocation Date: Friday, July 22, 2016 3:01:56 PM MDT America/Denver
                Extensions:
                    Identifier: Revocation Reason - 2.5.29.21
                        Critical: no
                        Reason: Certificate_Hold
        Extensions:
            Identifier: Authority Key Identifier - 2.5.29.35
                Critical: no
                Key Identifier:
                    BB:36:98:5D:65:CB:88:E0:87:23:37:6F:5B:F7:AF:8B:
                    8A:EB:BA:B5
            Identifier: CRL Number - 2.5.29.20
                Critical: no
                Number: 10
        Signature:
            Algorithm: SHA256withRSA - 1.2.840.113549.1.1.11
            Signature:
                A3:9F:0A:5D:2D:55:38:B2:0D:C6:9C:E1:50:88:7A:41:
                B9:6B:29:7A:FC:E3:D2:9A:07:42:CF:5B:E1:4B:F5:EE:
                8D:0C:0B:88:AB:7E:D9:11:CD:10:55:83:02:EF:75:27:
                79:58:51:14:49:B8:56:56:38:52:E2:8E:30:F3:80:57:
                76:6B:62:FF:E5:01:4B:E9:8E:54:DC:62:2A:AB:37:67:
                B1:71:72:0F:A6:0A:C8:87:48:35:93:DE:28:AA:65:5A:
                A5:3C:C3:8A:0F:32:10:1A:3E:17:78:0A:D2:04:AD:77:
                AA:33:5D:C0:AC:D5:25:99:57:61:FC:33:74:3B:72:54:
                A0:49:72:81:13:5C:66:77:8A:B5:3C:45:35:E8:EC:E8:
                4E:EB:12:83:01:A1:E3:CD:55:72:9E:37:C6:74:80:DA:
                6A:F9:75:7E:3C:17:16:E7:F0:64:17:A2:5E:6D:73:CB:
                21:52:BC:35:1D:9F:2E:06:DB:D3:53:98:CD:BD:E3:98:
                C2:11:27:91:EC:35:76:AB:52:C7:1B:FB:F6:AF:30:8E:
                29:35:9D:3D:D5:D3:3C:28:9C:B3:4E:4A:CD:40:38:4B:
                A5:FE:9A:9D:39:D7:09:21:C8:93:36:2D:8A:55:EA:52:
                95:E8:98:9C:CC:72:EB:81:FE:50:72:A7:F0:4A:44:D7
```

## SEE ALSO

**PrettyPrintCert(1)**, **pki(1)**

## AUTHORS

Matthew Harmsen &lt;mharmsen@redhat.com&gt;.

## COPYRIGHT

Copyright (c) 2016 Red Hat, Inc.
This is licensed under the GNU General Public License, version 2 (GPLv2).
A copy of this license is available at http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.


base/java-tools/src/com/netscape/cmstools/PrettyPrintCrl.java
// --- BEGIN COPYRIGHT BLOCK ---
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//
// (C) 2007 Red Hat, Inc.
// All rights reserved.
// --- END COPYRIGHT BLOCK ---
package com.netscape.cmstools;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.security.cert.CRLException;
import java.security.cert.CertificateException;
import java.util.Locale;

import netscape.security.util.CrlPrettyPrint;
import netscape.security.x509.DeltaCRLIndicatorExtension;
import netscape.security.x509.HoldInstructionExtension;
import netscape.security.x509.InvalidityDateExtension;
import netscape.security.x509.IssuingDistributionPointExtension;
import netscape.security.x509.OIDMap;
import netscape.security.x509.X509CRLImpl;
import netscape.security.x509.X509ExtensionException;

import com.netscape.cmsutil.util.Utils;

/**
 * The PrettyPrintCrl class is a utility program designed to "pretty print"
 * a CRL. It assumes that the name of a data file is passed to the
 * program via the command line, and that the contents contain a CRL
 * encoded in an ASCII BASE 64 format. Note that the data file may contain
 * an optional "-----BEGIN" header and/or an optional "-----END" trailer.
 *
 * <P>
 * The program may be invoked as follows:
 *
 * <PRE>
 *
 *      PrettyPrintCrl &lt;input filename&gt; [output filename]
 *
 *      NOTE:  &lt;input filename&gt;   must contain an ASCII
 *                                BASE 64 encoded CRL
 *
 *             &lt;output filename&gt;  contains a CRL displayed
 *                                in a "pretty print" ASCII format
 * </PRE>
 *
 * @version $Revision$, $Date$
 */

public class PrettyPrintCrl {
    // Define constants
    public static final int ARGC = 2;
    public static final String HEADER = "-----BEGIN";
    public static final String TRAILER = "-----END";

    public static void main(String argv[]) {

        BufferedReader inputCrl = null;
        String encodedBASE64CrlChunk = "";
        StringBuffer encodedBASE64Crl = new StringBuffer();

        byte decodedBASE64Crl[] = null;
        X509CRLImpl crl = null;
        Locale aLocale = null;
        CrlPrettyPrint CrlDetails = null;
        String pp;
        FileOutputStream outputCrl = null;

        // (1) Check that at least one argument was submitted to the program
        if ((argv.length < 1) || (argv.length > ARGC)) {
            System.out.println("Usage:  PrettyPrintCrl " +
                    "<input filename> " +
                    "[output filename]");
            return;
        }

        try {
            OIDMap.addAttribute(DeltaCRLIndicatorExtension.class.getName(),
                    DeltaCRLIndicatorExtension.OID,
                    DeltaCRLIndicatorExtension.NAME);
        } catch (CertificateException e) {
        }
        try {
            OIDMap.addAttribute(HoldInstructionExtension.class.getName(),
                    HoldInstructionExtension.OID,
                    HoldInstructionExtension.NAME);
        } catch (CertificateException e) {
        }
        try {
            OIDMap.addAttribute(InvalidityDateExtension.class.getName(),
                    InvalidityDateExtension.OID,
                    InvalidityDateExtension.NAME);
        } catch (CertificateException e) {
        }
        try {
            OIDMap.addAttribute(IssuingDistributionPointExtension.class.getName(),
                    IssuingDistributionPointExtension.OID,
                    IssuingDistributionPointExtension.NAME);
        } catch (CertificateException e) {
        }

        // (2) Create a DataInputStream() object to the BASE 64
        //     encoded CRL contained within the file
        //     specified on the command line
        try {
            inputCrl = new BufferedReader(new InputStreamReader(
                            new BufferedInputStream(
                                    new FileInputStream(
                                            argv[0]))));
        } catch (FileNotFoundException e) {
            System.out.println("PrettyPrintCrl():  can''t find file " +
                    argv[0] + ":\n" + e);
            return;
        }

        // (3) Read the entire contents of the specified BASE 64 encoded
        //     CRL into a String() object throwing away any
        //     headers beginning with HEADER and any trailers beginning
        //     with TRAILER
        try {
            while ((encodedBASE64CrlChunk = inputCrl.readLine()) != null) {
                if (!(encodedBASE64CrlChunk.startsWith(HEADER)) &&
                        !(encodedBASE64CrlChunk.startsWith(TRAILER))) {
                    encodedBASE64Crl.append(encodedBASE64CrlChunk.trim());
                }
            }
        } catch (IOException e) {
            System.out.println("PrettyPrintCrl():  Unexpected BASE64 " +
                    "encoded error encountered in readLine():\n" +
                    e);
        }

        // (4) Close the DataInputStream() object
        try {
            inputCrl.close();
        } catch (IOException e) {
            System.out.println("PrettyPrintCrl():  Unexpected BASE64 " +
                    "encoded error encountered in close():\n" + e);
        }

        // (5) Decode the ASCII BASE 64 CRL enclosed in the
        //     String() object into a BINARY BASE 64 byte[] object

        decodedBASE64Crl = Utils.base64decode(encodedBASE64Crl.toString());

        // (6) Create an X509CRLImpl() object from the BINARY BASE 64
        //     byte[] object
        try {
            crl = new X509CRLImpl(decodedBASE64Crl);
        } catch (CRLException e) {
            System.out.println("PrettyPrintCrl():  Error encountered " +
                    "on parsing and initialization errors:\n" + e);
        } catch (X509ExtensionException e) {
            System.out.println("PrettyPrintCrl():  Error encountered " +
                    "on parsing and initialization errors:\n" + e);
        }

        // (7) For this utility, always specify the default Locale
        aLocale = Locale.getDefault();

        // (8) Create a CrlPrettyPrint() object
        CrlDetails = new CrlPrettyPrint(crl);

        // (9) Convert the CrlPrettyPrint() object into a String() object
        pp = CrlDetails.toString(aLocale);

        // (10) Finally, "pretty print" the actual CRL to the console
        //      unless an output file has been specified
        if (argv.length != ARGC) {
            System.out.println(pp);
        } else {
            try {
                outputCrl = new FileOutputStream(argv[1]);
            } catch (IOException e) {
                System.out.println("PrettyPrintCrl():  unable to open file " +
                        argv[1] + " for writing:\n" + e);
                return;
            }

            try {
                outputCrl.write(pp.getBytes());
            } catch (IOException e) {
                System.out.println("PrettyPrintCrl():  I/O error " +
                        "encountered during write():\n" +
                        e);
            }

            try {
                outputCrl.close();
            } catch (IOException e) {
                System.out.println("PrettyPrintCrl():  Unexpected error " +
                        "encountered while attempting to close() " +
                        argv[1] + ":\n" + e);
            }
        }
    }
}
